core/src/main/java/org/infinispan/persistence/file/SingleFileStore.java

```diff
@@ -949,6 +1223,72 @@ public SingleFileStoreConfiguration getConfiguration() {
       return configuration;
    }
 
+   @Override
+   public CompletionStage<Void> addSegments(IntSet segments) {
+      if (!segmented) {
```

Comment:
`PersistenceManagerImpl.shouldInvokeSegmentMethods()` doesn't know anything about the store's configuration, it only checks if the store has the SEGMENTABLE characteristic.

Reply:
Which is updated in https://github.com/infinispan/infinispan/blob/main/core/src/main/java/org/infinispan/persistence/manager/PersistenceManagerImpl.java#L266

Reply:
Sorry, but that looks like a hack to me :) The way I read it, the characteristic is called "segmentable", so its presence means the store could be segmented, but it doesn't have to be segmented.

If SEGMENTABLE really means SEGMENTED, then I guess I could change the `characteristics()` implementation to return SEGMENTABLE only when the configuration is segmented. The NonBlockingStore javadoc says different instances of the store could return different characteristics, as long as the characteristics don't change after start.

If not, then it would be better IMO to keep a separate `segmented` flag in `StoreStatus` and to mention explicitly in the NonBlockingStore javadoc which methods are only called when the configuration is segmented.

Reply:
> Sorry, but that looks like a hack to me :) The way I read it, the characteristic is called "segmentable", so its presence means the store could be segmented, but it doesn't have to be segmented.

That is exactly what it does. The store status stores whether it actually is or not at runtime.

> If SEGMENTABLE really means SEGMENTED, then I guess I could change the `characteristics()` implementation to return SEGMENTABLE only when the configuration is segmented.

It doesn't mean that for the store. You are thinking solely from the store perspective. The store status just resuses it for its own purpose is all.

> The NonBlockingStore javadoc says different instances of the store could return different characteristics, as long as the characteristics don't change after start.

Yeah, that is perfectly fine. It works just fine either way.

> If not, then it would be better IMO to keep a separate `segmented` flag in `StoreStatus` and to mention explicitly in the NonBlockingStore javadoc which methods are only called when the configuration is segmented.

The characteristic is that flag on the status. It is already mentioned explicitly on the javadoc for every method. (e.g. https://github.com/infinispan/infinispan/blob/main/core/src/main/java/org/infinispan/persistence/spi/NonBlockingStore.java#L393)

Reply:
> It doesn't mean that for the store. You are thinking solely from the store perspective. The store status just resuses it for its own purpose is all.

And that's exactly what I'm arguing against: a characteristic SEGMENTABLE shouldn't have one meaning in the store and another meaning in PersistenceManagerImpl.

> It is already mentioned explicitly on the javadoc for every method. (e.g. https://github.com/infinispan/infinispan/blob/main/core/src/main/java/org/infinispan/persistence/spi/NonBlockingStore.java#L393)

My bad, I should have checked the javadoc for `addSegments()`.

In my defense, the not all the javadocs say the same thing. The `load()`, `delete()`, `write()`, `publishKeys()` and `publishEntries()` javadoc only says "When this is not set the provided {@code segments} parameter may be ignored.", implying that the segments parameter must not be ignored if the store reports that it supports segmentation but segmentation is disabled in the configuration.

And that is indeed how `DummyInMemoryStore` implements them (especially now that I added back the filter in `publishEntries()`).

Reply:
> And that's exactly what I'm arguing against: a characteristic SEGMENTABLE shouldn't have one meaning in the store and another meaning in PersistenceManagerImpl.

That's fine, it is personal preference imo. Note that this is done for more than just segmented as well and isn't as clear cut for some.



> In my defense, the not all the javadocs say the same thing. The `load()`, `delete()`, `write()`, `publishKeys()` and `publishEntries()` javadoc only says "When this is not set the provided {@code segments} parameter may be ignored.",

Yeah, because they are different. Note that publishKeys and publishEntries are not invoked if you don't have the BULK_READ characteristic https://github.com/infinispan/infinispan/blob/main/core/src/main/java/org/infinispan/persistence/spi/NonBlockingStore.java#L649

Same thing for write/delete with READ_ONLY https://github.com/infinispan/infinispan/blob/main/core/src/main/java/org/infinispan/persistence/spi/NonBlockingStore.java#L359

And same thing for load/contains with WRITE_ONLY
https://github.com/infinispan/infinispan/blob/main/core/src/main/java/org/infinispan/persistence/spi/NonBlockingStore.java#L286

> implying that the segments parameter must not be ignored if the store reports that it supports segmentation but segmentation is disabled in the configuration.

Sure, that is an oversight and please add in that it can be ignore if the store doesn't have SEGMENTABLE or StoreConfiguration#segmented is false.

> And that is indeed how DummyInMemoryStore implements them (especially now that I added back the filter in publishEntries()).

DO NOT add the filter method back in as I said on the other review comment.

Reply:
> Note that this is done for more than just segmented as well and isn't as clear cut for some.

Again IMO but the fact that it's less clear cut for other characteristics should be even more reason to have two different names for them, to make it clear when we're using one meaning and when we're using the other meaning. Otherwise we're just inviting confusion.

> Sure, that is an oversight and please add in that it can be ignore if the store doesn't have SEGMENTABLE or StoreConfiguration#segmented is false.

:+1: 

> DO NOT add the filter method back in as I said on the other review comment.

I'd rather add it back now that I realized I am using the segments parameter in  `SingleFileStore.publishEntries()` and `RocksDBStore.NonSegmentedRocksDBHandler.publishEntries()` is also using it, but ok :)

Reply:
@wburns I changed `addSegments()` and `removeSegments()` to throw `UnsupportedOperationException`, but FWIW RocksDBTest (which I used as a template for SingleFileStoreTest) still tests that `addSegments` and `removeSegments` work when segmentation is disabled in the configuration. I haven't checked the other stores, but it seems pretty clear I wasn't the only one who didn't read the javadoc of `addSegments` and `removeSegments` :)